pilot/pkg/networking/util/util.go

```diff
@@ -758,59 +758,92 @@ func MaybeBuildStatefulSessionFilterConfig(svc *model.Service) *statefulsession.
-		mergedPolicy.Tls = original.Tls
-		mergedPolicy.Tunnel = original.Tunnel
-		mergedPolicy.ProxyProtocol = original.ProxyProtocol
+	// settings specified at the destination-level will not be inherited when
+	// overridden by port-level settings, i.e. default values will be applied
+	// to fields omitted in port-level traffic policies.
+	ret := &networking.TrafficPolicy{}
+	ret.ConnectionPool = portTrafficPolicy.ConnectionPool
+	ret.LoadBalancer = portTrafficPolicy.LoadBalancer
+	ret.OutlierDetection = portTrafficPolicy.OutlierDetection
+	ret.Tls = portTrafficPolicy.Tls
+	return ret, true
+}
+
+// MergeSubsetTrafficPolicy merges the destination and subset level traffic policy for the given port.
+func MergeSubsetTrafficPolicy(original, subsetPolicy *networking.TrafficPolicy, port *model.Port) *networking.TrafficPolicy {
+	// First get DR port level traffic policy
+	original, _ = GetPortLevelTrafficPolicy(original, port)
+	if subsetPolicy == nil {
+		return original
+	}
+	subsetPolicy, hasPortLevel := GetPortLevelTrafficPolicy(subsetPolicy, port)
+	if original == nil {
+		return subsetPolicy
 	}
 
+	// merge DR with subset traffic policy
+	mergedPolicy := CloneTrafficPolicy(original)
+
 	// Override with subset values.
-	if subsetPolicy.ConnectionPool != nil {
+
+	// settings specified at the destination-level will not be inherited when
+	// overridden by port-level settings, i.e. default values will be applied
+	// to fields omitted in port-level traffic policies.
+	if subsetPolicy.ConnectionPool != nil || hasPortLevel {
 		mergedPolicy.ConnectionPool = subsetPolicy.ConnectionPool
 	}
-	if subsetPolicy.OutlierDetection != nil {
+	if subsetPolicy.OutlierDetection != nil || hasPortLevel {
 		mergedPolicy.OutlierDetection = subsetPolicy.OutlierDetection
 	}
-	if subsetPolicy.LoadBalancer != nil {
+	if subsetPolicy.LoadBalancer != nil || hasPortLevel {
 		mergedPolicy.LoadBalancer = subsetPolicy.LoadBalancer
 	}
-	if subsetPolicy.Tls != nil {
+	if subsetPolicy.Tls != nil || hasPortLevel {
 		mergedPolicy.Tls = subsetPolicy.Tls
 	}
+
 	if subsetPolicy.Tunnel != nil {
 		mergedPolicy.Tunnel = subsetPolicy.Tunnel
 	}
 	if subsetPolicy.ProxyProtocol != nil {
 		mergedPolicy.ProxyProtocol = subsetPolicy.ProxyProtocol
 	}
+	return mergedPolicy
+}
 
-	// Check if port level overrides exist, if yes override with them.
-	if port != nil {
-		for _, p := range subsetPolicy.PortLevelSettings {
-			if p.Port != nil && uint32(port.Port) == p.Port.Number {
-				// per the docs, port level policies do not inherit and instead to defaults if not provided
-				mergedPolicy.ConnectionPool = p.ConnectionPool
-				mergedPolicy.OutlierDetection = p.OutlierDetection
-				mergedPolicy.LoadBalancer = p.LoadBalancer
-				mergedPolicy.Tls = p.Tls
-				break
-			}
-		}
+// Shallowcopy a traffic policy, portLevelSettings are ignorred.
+func CloneTrafficPolicy(original *networking.TrafficPolicy) *networking.TrafficPolicy {
```

Comment:
Can you add a fuzz test for this to ensure the fields are not missed in future?

Reply:
It is not copying call fields